tokio/src/runtime/time/mod.rs

```diff
@@ -324,23 +323,53 @@ impl Handle {
-            if let Some(waker) = unsafe { entry.fire(Ok(())) } {
-                waker_list.push(waker);
-
-                if !waker_list.can_push() {
-                    // Wake a batch of wakers. To avoid deadlock, we must do this with the lock temporarily dropped.
-                    drop(lock);
-
-                    waker_list.wake_all();
-
-                    lock = self.inner.lock_sharded_wheel(id);
+        while let Some(expiration) = lock.poll(now) {
+            lock.set_elapsed(expiration.deadline);
+            // It is critical for `GuardedLinkedList` safety that the guard node is
+            // pinned in memory and is not dropped until the guarded list is dropped.
+            let guard = TimerShared::new(id);
+            pin!(guard);
+            let guard_handle = guard.as_ref().get_ref().handle();
+
+            // * This list will be still guarded by the lock of the Wheel with the specefied id.
+            //   `EntryWaitersList` wrapper makes sure we hold the lock to modify it.
+            // * This wrapper will empty the list on drop. It is critical for safety
+            //   that we will not leave any list entry with a pointer to the local
+            //   guard node after this function returns / panics.
+            // Safety: The `TimerShared` inside this `TimerHandle` is pinned in the memory.
+            let mut list = unsafe { lock.get_waiters_list(&expiration, guard_handle, id, self) };
+
+            while let Some(entry) = list.pop_back_locked(&mut lock) {
+                let deadline = expiration.deadline;
+                // Try to expire the entry; this is cheap (doesn't synchronize) if
+                // the timer is not expired, and updates cached_when.
+                match unsafe { entry.mark_firing(deadline) } {
+                    Ok(()) => {
+                        // Entry was expired.
+                        // SAFETY: We hold the driver lock, and just removed the entry from any linked lists.
+                        if let Some(waker) = unsafe { entry.fire(Ok(())) } {
+                            waker_list.push(waker);
+
+                            if !waker_list.can_push() {
+                                // Wake a batch of wakers. To avoid deadlock,
+                                // we must do this with the lock temporarily dropped.
+                                drop(lock);
+                                waker_list.wake_all();
+
+                                lock = self.inner.lock_sharded_wheel(id);
+                            }
+                        }
+                    }
+                    Err(state) if state == STATE_DEREGISTERED => {}
+                    Err(state) => {
+                        // Safety: This Entry has not expired.
+                        unsafe { lock.reinsert_entry(entry, deadline, state) };
+                    }
                 }
             }
+            lock.occupied_bit_maintain(&expiration);
```

Comment:
The old code doesn't call this. Why is it necessary now?

Reply:
The reason I add `occupied_bit_maintain` now is that in the past, we would use `take_entries` to take all items from the linked list at once and then update the occupied bit. But now, we need to traverse the items one by one, which means we need to update the occupied bit after the traversal is completed. I think the difference here is that in the current version, the linked list is not necessarily empty, so I add this method.

Reply:
Would it be accurate to say that this is an operation that must be called when we are done with `list`? Would it make sense for it to be part of the destructor of `EntryWaitersList`?

Reply:
Does it need to be called before `drop(lock)` in `if !waker_list.can_push() {`?

Reply:
Calling before `drop(lock)` in `if !waker_list.can_push() {` makes sense to me. That is a good idea.

Reply:
What about `set_elapsed`?

I would really like to see a loom test where a timer is inserted during the `if !waker_list.can_push() { ... }` block. It's hard for me to tell whether it's correct.

I think you can do the loom test along these lines:

1. Start a runtime with paused time.
2. Create and poll [32](https://github.com/tokio-rs/tokio/blob/955ec31d95a936949585247b334cd8c6629803d6/tokio/src/util/wake_list.rs#L5) timers with the timer.
3. Spawn a thread. The background thread will create a timer, register it, and wait for it to complete.
4. In parallel with that wait for the 32 timers in order.
5. Join the background thread.

Spawning the thread after registering the 32 timers should reduce the size of the loom test, since the region with more than 1 thread will be shorter. As long as the `if !waker_list.can_push() { ... }` block happens in the more-than-1-thread region of the test, it should work.